Restructure recorder documentation to contain all properties.

jougs opened the pull request:
This PR is a replacement for #1819 and in essence adds the complete table of recorder properties to each recorder by including a documentation snippet extracted from `nestkernel/recording_device.h`, where the `RecordingBackend` base class is defined.

On the infrastructure side, I have added a line of code to handle a new magic `NOINDEX` keyword to the user doc extractor, which excludes the generated documentation from any indices in the model directory, and replaced the direct inclusion of recorder and recording backend documentation into the guide on recording from simulations by links to get the size of that page down.

This PR implements the changes discussed amongst the author and the suggested reviewers on December 12, 2020.

@jhnnsnk: please feel free to send me the example for switching recording backends you have presented at that meeting as a PR against my branch.

sarakonradi commented on nestkernel/recording_device.h:
```diff
@@ -42,6 +42,62 @@
+Recording time window
++++++++++++++++++++++
+
+The time span during which the recorder actually records, can be
+specified using the properties ``start`` and ``stop``. These define
```
```suggestion
The time span during which the recorder actually records can be
specified using the properties ``start`` and ``stop``. These define
```

sarakonradi replied:
Can we perhaps replace the word "actually" with "actively"?

jougs replied:
Yes.
```suggestion
The time span during which the recorder actively records, can be
specified using the properties ``start`` and ``stop``. These define
```

sarakonradi commented on nestkernel/recording_device.h:
```diff
@@ -42,6 +42,62 @@
+ stop
+   A floating point number (default: `infinity`) specifying the
+   deactication time in ms, relative to `origin`. The value of `stop`
+   must be greater than or equal to `start`
```
```suggestion
   must be greater than or equal to `start`.
```

sarakonradi reviewed (changes requested):
Thanks for restructuring the documentation! Just added a few comments.

jougs commented on doc/guides/recording_from_simulations.rst:
```diff
@@ -35,87 +24,35 @@ temporal aspects of the simulation loop.
-   A string (default: `"memory"`) containing the name of the recording
-   backend where to write data to. An empty string turns all recording
-   of individual events off.
+- **Collectors** collect events sent to them. Neurons are connected to
+  collectors and the collector collects the events emitted by the
+  neurons connected to it.
```
Good idea.
```suggestion
- **Collectors** gather events sent to them. Neurons are connected to
  collectors and the collector gathers the events emitted by the
  neurons connected to it.
```

terhorstd commented on doc/guides/connection_management.rst:
```diff
@@ -1,4 +1,4 @@
-.. _connection_mgnt:
+.. _connection_management:
```
Is this required at all? I thought Sphinx automatically creates anchors for titles?

jougs replied:
I think it does. However, I'd rather rely on a static name for references, instead of one that changes when the title changes...

terhorstd commented on models/spike_recorder.h:
```diff
@@ -64,6 +66,8 @@ spike creation rather than that of their arrival.
 The call to ``Connect`` will fail if the connection direction is
 reversed (i.e., connecting *sr* to *neurons*).
 
+.. include:: ../models/recording_device.rst
+
 EndUserDocs */
```
Add a See-also section to actually refer to other recorders
```suggestion
.. include:: ../models/recording_device.rst

See also
++++++++

EndUserDocs */
```

terhorstd commented on models/weight_recorder.h:
```diff
@@ -71,6 +73,8 @@ synapses that fulfill the given criteria.
 
    >>> nest.Connect(pre, post, syn_spec="stdp_synapse_rec")
 
+.. include:: ../models/recording_device.rst
+
 EndUserDocs */
```
Add a See-also section to actually refer to other recorders
```suggestion
.. include:: ../models/recording_device.rst

See also
++++++++

EndUserDocs */
```

terhorstd commented on models/multimeter.h:
```diff
@@ -110,6 +112,19 @@ fail if carried out in the wrong direction, i.e., trying to connect the
+ interval
+   A float (default: `1.0`) specifying the interval in ms, at which
+   data is collected from the nodes, the multimeter is connected to.
+
 EndUserDocs */
```
Add a See-also section to actually refer to other recorders
```suggestion

See also
++++++++

EndUserDocs */
```

terhorstd commented on doc/extractor_userdocs.py:
```diff
@@ -372,6 +372,9 @@ def mkitem(t):
 
     for tags, items in sorted(hierarchy.items()):
+
+        if "NOINDEX" in tags:
+            continue
```
Just checked. Currently the included files do not have a "See also" section, so they do not link to the "NOINDEX" index. So all is fine here, until we decide to automatically add "See also" sections where missing (which we probably shouldn't anyway)

jougs replied:
Thanks for the clarification.

terhorstd reviewed (changes requested):
Thanks for the restructuring, it improves the recorders docs significantly!
Two small comments:
1. Maybe add a sentence to the "Recorder Properties" section, that this is what can be given in the `params` argument. There is no clear link between parameters and properties on the semantic level. An inline example including a parameter would probably be enough; details are in the "recording from simulations guide" anyway.

2. Add the "See also", sections in the relevant places to have the links to keyword indices, e.g. where suggested.
 
With those small additions it would be fine for me.

jougs commented:
@terhorstd: Thanks for your review. Can you please have another quick look?

@jhnnsnk: Any change to include your NumPy example in this one? If not, we can also do that in another PR.

jougs commented:
@terhorstd: here's an observation related to this issue, I'd like to have your opinion on:

I am still convinced that we need the compositionality of documentation files via the 'NOINDEX' tag and `.. include::` that is added in this PR. However, this now bites us a bit in the context of #1926, where the model documentation now just shows a useless include instead of the model properties. Do you think it would be smart to dissolve includes (worst case manually) before actually letting Sphinx create HTML from the RST? This way the resulting RST file would be more self-contained and completee and as a bonus there simply wouldn't be anything to be added or skipped from any index. Thanks!

jougs commented:
@clinssen created #1930 to describe the naming issue more concisely and suggest a solution also for the problem described [above](https://github.com/nest/nest-simulator/pull/1899#issuecomment-780485133).

clinssen commented:
Potentially, we could render the rST files (which contain the include directives) to another plaintext format, such as MarkDown or man-page (which should be supported by docutils). This could either happen offline (during installation) or online (when the "help" command is invoked).

jougs commented:
@clinssen: I have just tried several different tools (some based on [Docutils](https://docutils.sourceforge.io/), some on [pandoc](https://pandoc.org/)) for converting our `.rst` files to MarkDown. They alll failed. Either they replaced unknown roles (e.g. links to other documentation pages [`:doc:`]) by a `Not supported` error message, or they just skipped pertaining text blocks completely. Totally unacceptable. :-(

Then I tried [something different](https://stackoverflow.com/a/47130801). Sphinx apparently has a Markdown builder since some time and that somehow works better, but totally messes up any tables in the `.rst` file and also skips references and foot notes and at least in some cases equations seem to be missing from the output.

As Sphinx was my best bet in this game, I'm afraid that our best choice at this point is really to manually replace the includes and present the source `.rst`, as that is the only one guaranteed to be complete.

jougs commented:
@terhorstd, I know you're busy, but I'd really very much appreciate your pandoc-foo expertise in these matters. Come join us on the dark dirty scripting side ;-)

heplesser commented:
@jougs After discussions in recent days, what are the plans now for including material in the online help?

jougs commented:
Here's a summary of my recent activities:
* I've merged upstream master in 0e380e0 to resolve conflicts
* I've added a new PyNEST example in d590e7aa4 that shows how to use different recording backends;
  kudos to @jhnnsnk for the initial version
* regarding the issue with the includes, I have added a simple include-resolving script in 7c56528
   @terhorstd: could you please have a quick look?

jougs commented:
Could all of you please re-review? I think I have all points addressed.

jougs commented:
Thanks!

sarakonradi commented on nestkernel/recording_device.h:
```diff
@@ -42,6 +42,62 @@
+Recording time window
++++++++++++++++++++++
+
+The time span during which the recorder actively records, can be
```
```suggestion
The time span during which the recorder actively records can be
```

sarakonradi commented on pynest/examples/recording_demo.py:
```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+#
+# recording_demo.py
```
We should add this to ``doc/userdoc/examples/index.rst``.

jougs replied:
Good idea! Done in af7a69bfc.

sarakonradi reviewed (changes requested):
Many thanks @jougs! Here are two minor suggestions.

jougs commented:
@sarakonradi: I've just addressed all your concerns.

jougs commented:
@terhorstd: could you please have a quick look at the include-resolving script added in 7c56528? Thanks!

ackurth commented on doc/userdoc/guides/recording_from_simulations.rst:
```diff
@@ -35,87 +24,35 @@ temporal aspects of the simulation loop.
-After a recording device has sampled or collected data, the *recording backends* are
-responsible for how the data are processed.
-
+After a recording device has sampled or collected data, the data is
```
Minor point: Maybe `collected and sampled data` is better than `sampled and collected data` to preserve the order in which the groups of recording devices are introduced.

jougs replied:
```suggestion
After a recording device has collected or sampled data, the data is
```

ackurth commented on nestkernel/recording_device.h:
```diff
@@ -42,6 +42,62 @@
+
+The time span during which the recorder actively records can be
+specified using the properties ``start`` and ``stop``. These define
+the activation period of the device in ms. An additional property
```
If find `activation period` to be a bit ambiguous in this case since to my knowledge in a Neuroscientific context it's usually related to neural activation. Maybe `recording period` is better?

jougs replied:
```suggestion
the recording period of the device in ms. An additional property
```

ackurth commented on nestkernel/recording_device.h:
```diff
@@ -42,6 +42,62 @@
+
+ stop
+   A floating point number (default: `infinity`) specifying the
+   deactication time in ms, relative to `origin`. The value of `stop`
```
```suggestion
   deactivation time in ms, relative to `origin`. The value of `stop`
```

ackurth reviewed (commented):
I think the documentation is very comprehensible, just added some minor comments.

jougs reviewed (commented):
Just adding suggestions to address @ackurth's review comments.

clinssen reviewed (approved):
Looks good!

terhorstd reviewed (approved):
The include script is simple but efficient.
Looks good to me now.
